File: api/utils/worldCrypto.js
'use strict';
const crypto = require('crypto');

/**
 * AES-256-GCM world content encryption utilities.
 * Uses Node.js built-in crypto — no third-party deps.
 */

/**
 * Derive a per-session key from the world's master key using HMAC-SHA256.
 * @param {Buffer|string} masterKey - 32-byte master key (hex string or Buffer)
 * @param {string} sessionId
 * @returns {Buffer} 32-byte session key
 */
function deriveSessionKey(masterKey, sessionId) {
  const key = typeof masterKey === 'string' ? Buffer.from(masterKey, 'hex') : masterKey;
  return crypto.createHmac('sha256', key).update(sessionId).digest();
}

/**
 * Encrypt a node/scene object with AES-256-GCM.
 * @param {object} plaintext
 * @param {Buffer} key - 32-byte session key
 * @returns {string} "base64(enc)~base64(iv)~base64(authTag)"
 */
function encryptNode(plaintext, key) {
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
  const enc = Buffer.concat([
    cipher.update(JSON.stringify(plaintext), 'utf8'),
    cipher.final(),
  ]);
  const authTag = cipher.getAuthTag();
  return [enc, iv, authTag].map(b => b.toString('base64')).join('~');
}

/**
 * Decrypt a node/scene string produced by encryptNode.
 * @param {string} ciphertext
 * @param {Buffer} key - 32-byte session key
 * @returns {object}
 */
function decryptNode(ciphertext, key) {
  const [enc, iv, authTag] = ciphertext.split('~').map(s => Buffer.from(s, 'base64'));
  const decipher = crypto.createDecipheriv('aes-256-gcm', key, iv);
  decipher.setAuthTag(authTag);
  const plain = Buffer.concat([decipher.update(enc), decipher.final()]).toString('utf8');
  return JSON.parse(plain);
}

module.exports = { deriveSessionKey, encryptNode, decryptNode };

File: api/utils/world_crypto.py
"""
AES-256-GCM 加密工具 — Python 版
用于服务端加密/解密关卡场景数据，按回合下发给客户端。
"""
import hashlib
import hmac
import json
import os
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def derive_session_key(master_key: bytes, session_id: str) -> bytes:
    """用 HMAC-SHA256 从 master_key 派生 session 级密钥"""
    return hmac.new(master_key, session_id.encode(), hashlib.sha256).digest()


def generate_master_key() -> bytes:
    """生成 32 字节随机 master key"""
    return os.urandom(32)


def encrypt_node(plaintext: dict, key: bytes) -> str:
    """
    AES-256-GCM 加密场景节点。
    返回格式: base64(nonce) ~ base64(ciphertext+tag)
    """
    import base64
    nonce = os.urandom(12)
    aesgcm = AESGCM(key)
    ct = aesgcm.encrypt(nonce, json.dumps(plaintext, ensure_ascii=False).encode(), None)
    return base64.b64encode(nonce).decode() + '~' + base64.b64encode(ct).decode()


def decrypt_node(ciphertext: str, key: bytes) -> dict:
    """解密 encrypt_node 产出的密文"""
    import base64
    parts = ciphertext.split('~')
    nonce = base64.b64decode(parts[0])
    ct = base64.b64decode(parts[1])
    aesgcm = AESGCM(key)
    plain = aesgcm.decrypt(nonce, ct, None)
    return json.loads(plain.decode())
